### PR TITLE
Refactor EstablishClaimEmail

### PR DESCRIPTION
**Pull request by joofsh**
- Refactor Dispatch by moving logic out of EstablishClaim.jsx and into EstablishClaimEmail or the utility file

connects #2723

Testing:
- [x] Manually running through dispatch
- [x] All existing tests pass. I know they cover this body of code because they did fail at one point mid-refactor.
- [x] New tests added!

### CODE REVIEW

File: client/app/containers/EstablishClaimPage/EstablishClaimEmail.jsx [hunk 1]
```diff
@@ -196,12 +153,43 @@ EstablishClaimEmail.propTypes = {
...
+  specialIssues: state.specialIssues,
+  loading: state.establishClaim.loading
 });
 
-const mapDispatchToProps = (dispatch) => ({
+const mapDispatchToProps = (dispatch, ownProps) => ({
   handleToggleCancelTaskModal: () => {
     dispatch({ type: Constants.TOGGLE_CANCEL_TASK_MODAL });
+  },
+  handleEmailSubmit: () => {
```
Comment (joofsh) [inline client/app/containers/EstablishClaimPage/EstablishClaimEmail.jsx]: moved here from EstablishClaim.jsx

File: client/app/establishClaim/reducers/store.js [hunk 2]
```diff
@@ -7,6 +7,9 @@ import establishClaimFormReducer,
   { getEstablishClaimFormInitialState } from './establishClaimForm';
 import ConfigUtil from '../../util/ConfigUtil';
 
+// eslint-disable-next-line no-underscore-dangle
+const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
```
Comment (joofsh) [inline client/app/establishClaim/reducers/store.js]: not directly related to this PR. But adding redux dev tools to Dispatch

File: client/test/karma/establishClaim/util/index-test.js [hunk 3]
```diff
@@ -76,4 +80,66 @@ context('establishClaimUtil', () => {
       expect(result).to.eql(modifiers);
     });
   });
+
+  context('.getSpecialIssuesRegionalOfficeCode', () => {
```
Comment (joofsh) [inline client/test/karma/establishClaim/util/index-test.js]: yay more unit tests!

File: client/app/containers/EstablishClaimPage/EstablishClaim.jsx [hunk 4]
```diff
@@ -629,15 +524,27 @@ export default class EstablishClaim extends BaseForm {
...
-            loading={this.state.loading}
             appeal={this.props.task.appeal}
-            handleEmailSubmit={this.handleEmailPageSubmit}
-            handleNoEmailSubmit={this.handleNoEmailPageSubmit}
-            regionalOffice={this.getSpecialIssuesRegionalOffice()}
-            regionalOfficeEmail={this.getSpecialIssuesEmail()}
-            specialIssues={specialIssues}
+            handleAlertClear={this.props.handleAlertClear}
+            handleAlert={this.props.handleAlert}
+            regionalOfficeEmail={
```
Comment (joofsh) [inline client/app/containers/EstablishClaimPage/EstablishClaim.jsx]: these two props are kind of gross, but I think it's a good intermediate step until `this.state.specialIssuesEmail` and `this.state.specialIssuesRegionalOffice` are converted to Redux. I wanted to try to keep this PR to ~200 line diff

File: client/app/containers/EstablishClaimPage/EstablishClaim.jsx [hunk 5]
```diff
@@ -361,59 +363,6 @@ export default class EstablishClaim extends BaseForm {
...
-        'Error',
-        'There was an error while completing the task. Please try again later'
-        );
-        this.setState({
-          loading: false
-        });
-      });
-  };
-
-  handleNoEmailPageSubmit = () => {
```
Comment (sharonwarner) [inline client/app/containers/EstablishClaimPage/EstablishClaim.jsx]: Yes, that's correct.

File: client/app/establishClaim/reducers/store.js (hunk 2 above)
Comment (sharonwarner) [inline client/app/establishClaim/reducers/store.js]: 👍

File: client/app/containers/EstablishClaimPage/EstablishClaim.jsx (hunk 4 above)
Comment (NickHeiner) [inline client/app/containers/EstablishClaimPage/EstablishClaim.jsx]: image

:smile:

File: client/app/establishClaim/reducers/index.js [hunk 6]
```diff
@@ -34,6 +35,11 @@ export const establishClaim = function(state = getEstablishClaimInitialState(),
       isValidating: false,
       isShowingCancelModal: !state.isShowingCancelModal
     };
+  case Constants.TRIGGER_LOADING:
+    return {
```
Comment (NickHeiner) [inline client/app/establishClaim/reducers/index.js]: It's fine if you don't want to do it in this PR, but I'd strongly recommend moving to `immutability-helper` eventually, like we do in Reader. With that, this looks like:

```js
return update(state, {
  loading: {
    $set: action.payload.value
  }
});
```

For a case like this, it looks pretty similar, but when your Redux state is more complex, it really simplifies things.

Also, for toggles, you can do cool stuff like:

```js
const toggleValue = (val) => !val;

// ...

case Constants.TOGGLE_CANCEL_TASK_MODAL:
  return update(state, {
    isShowingCancelModal: { $apply: toggleValue }
  });
```

File: client/app/establishClaim/reducers/store.js (hunk 2 above)
Comment (NickHeiner) [inline client/app/establishClaim/reducers/store.js]: :+1: 💯

File: client/app/establishClaim/util/index.js [hunk 7]
```diff
@@ -54,3 +55,52 @@ export const validModifiers = (endProducts) => {
...
+      regionalOfficeCities[regionalOfficeKey].city}, ${
+      regionalOfficeCities[regionalOfficeKey].state}`;
+};
+
+const getEmailFromConstant = (constant, regionalOfficeKey) => {
+  return ROUTING_INFORMATION.codeToEmailMapper[constant[regionalOfficeKey]];
+};
+
+export const getSpecialIssuesRegionalOfficeCode = (specialIssuesRegionalOffice, regionalOfficeKey) => {
+  if (specialIssuesRegionalOffice === 'PMC') {
```
Comment (NickHeiner) [inline client/app/establishClaim/util/index.js]: This could read nicer as a `switch` statement but w/e.

File: client/app/establishClaim/util/index.js [hunk 8]
```diff
@@ -54,3 +55,52 @@ export const validModifiers = (endProducts) => {
...
+  return specialIssuesRegionalOffice;
+
+};
+
+// This method returns a string version of the regional office code. So it takes "R081"
+// and returns a string "RO81 - Philadelphia Pension Center, PA"
+export const getSpecialIssuesRegionalOffice =
+  (specialIssuesRegionalOffice, regionalOfficeKey, regionalOfficeCities) => {
+    return getRegionalOfficeString(
+    getSpecialIssuesRegionalOfficeCode(specialIssuesRegionalOffice, regionalOfficeKey),
```
Comment (NickHeiner) [inline client/app/establishClaim/util/index.js]: This line should be indented past the start of the previous line.

File: client/test/karma/establishClaim/util/index-test.js [hunk 9]
```diff
@@ -76,4 +80,66 @@ context('establishClaimUtil', () => {
...
+    const regionalOfficeKey = 'RO18';
+
+    it('when a PMC special issue', () => {
+      result = getSpecialIssuesRegionalOfficeCode('PMC', regionalOfficeKey);
+      expect(result).to.eql('RO81');
```
Comment (NickHeiner) [inline client/test/karma/establishClaim/util/index-test.js]: I don't like using `eql` as a shorthand for `deep.equal`, since it's less readable / obvious. But you don't have to change it here if you don't want to.

And you don't need to use `deep.equal` for strings, fwiw.

File: client/test/karma/establishClaim/util/index-test.js [hunk 10]
```diff
@@ -76,4 +80,66 @@ context('establishClaimUtil', () => {
       expect(result).to.eql(modifiers);
     });
   });
+
+  context('.getSpecialIssuesRegionalOfficeCode', () => {
+    const regionalOfficeKey = 'RO18';
+
+    it('when a PMC special issue', () => {
+      result = getSpecialIssuesRegionalOfficeCode('PMC', regionalOfficeKey);
```
Comment (NickHeiner) [inline client/test/karma/establishClaim/util/index-test.js]: What is the point of having `result` be defined in a scope higher than this test?

File: client/app/establishClaim/util/index.js (hunk 7 above)
Comment (joofsh) [inline client/app/establishClaim/util/index.js]: There are definitely opportunities for other refactors/improvements. I'm just copy/pasting code to a more appropriate place. I tried not to change the fundamental code too much

File: client/app/establishClaim/reducers/index.js (hunk 6 above)
Comment (joofsh) [inline client/app/establishClaim/reducers/index.js]: Good catch. Did we collectively decide to move forward with using immutability-helper everywhere? If so I'll make this change

Also fyi this isn't a toggle (and you can do a toggle just as simply without immutable-helper).

File: client/app/containers/EstablishClaimPage/EstablishClaimEmail.jsx [hunk 11]
```diff
@@ -190,18 +142,46 @@ EstablishClaimEmail.propTypes = {
...
+      ownProps.appeal.regional_office_key
+    );
+
+    const data = ApiUtil.convertToSnakeCase({
+      emailRoId,
+      emailRecipient: ownProps.regionalOfficeEmail.join(', ')
+    });
+
+    return ApiUtil.post(`/dispatch/establish-claim/${ownProps.taskId}/email-complete`, { data }).
+      then(() => {
```
Comment (joofsh) [inline client/app/containers/EstablishClaimPage/EstablishClaimEmail.jsx]: sure it's not needed, but it improves readability. It creates visual consistency between the promise resolve and reject (both have curly brackets). Let characters is not fundamentally better :)

File: client/test/karma/establishClaim/util/index-test.js (hunk 9 above)
Comment (joofsh) [inline client/test/karma/establishClaim/util/index-test.js]: good call, changed 👍

File: client/test/karma/establishClaim/util/index-test.js (hunk 10 above)
Comment (joofsh) [inline client/test/karma/establishClaim/util/index-test.js]: it removes "magic numbers". The same regionalOfficeKey is used in all 3 tests

File: client/app/establishClaim/reducers/index.js (hunk 6 above)
Comment (joofsh) [inline client/app/establishClaim/reducers/index.js]: updated 👍

File: client/app/establishClaim/util/index.js (hunk 7 above)
Comment (joofsh) [inline client/app/establishClaim/util/index.js]: Thinking on this more....I don't see us refactoring Dispatch again, so might as well do it right this time! Updated to use switch statements

File: client/app/establishClaim/reducers/index.js (hunk 6 above)
Comment (NickHeiner) [inline client/app/establishClaim/reducers/index.js]: I do not know that we collectively decided to use `immutability-helper` everywhere.

I know that this isn't a toggle. I was referring to `TOGGLE_CANCEL_TASK_MODAL` above.

File: client/test/karma/establishClaim/util/index-test.js (hunk 10 above)
Comment (NickHeiner) [inline client/test/karma/establishClaim/util/index-test.js]: I think that `no-magic-numbers` is a silly lint rule for tests.

I don't see how it helps to have `result` defined in a higher scope. How is that related to the fact that the same `regionalOfficeKey` value is used?

File: client/app/establishClaim/util/index.js [hunk 12]
```diff
@@ -54,3 +55,53 @@ export const validModifiers = (endProducts) => {
...
+  if (!regionalOfficeKey) {
+    return null;
+  }
+
+  return `${regionalOfficeKey} - ${
+      regionalOfficeCities[regionalOfficeKey].city}, ${
+      regionalOfficeCities[regionalOfficeKey].state}`;
+};
+
+const getEmailFromConstant = (constant, regionalOfficeKey) => {
```
Comment (NickHeiner) [inline client/app/establishClaim/util/index.js]: I find this to be more readable, but w/e.

```js

const getEmailFromConstant = (constant, regionalOfficeKey) =>
  ROUTING_INFORMATION.codeToEmailMapper[constant[regionalOfficeKey]];
```

For me, less visual clutter means more readability. But this is subjective, so I'll stop haranguing about it. :smile:

**Comment by NickHeiner** [comment]
I've verified that the tests fail when the code is broken.

File: client/test/karma/establishClaim/util/index-test.js (hunk 10 above)
Comment (joofsh) [inline client/test/karma/establishClaim/util/index-test.js]: sorry I misread your original comment. Defining `result` in the higher scope isn't necessary, you are correct. I guess I find it annoying to do `const result` in every test scope? But subjective
